File: Android/app/src/main/kotlin/br/com/victorwads/goldenunicorn/crypt/NumericDecryptor.kt
package br.com.victorwads.goldenunicorn.crypt

import com.google.firebase.auth.FirebaseAuth
import java.math.BigInteger
import java.security.MessageDigest

object NumericDecryptor {
    private const val FLOAT_PRECISION = 1000
    private const val MASK_SIZE = 2 // 2 bits
    private val MASK = BigInteger.valueOf(0b11L)
    private val BOOLEAN_MASK = BigInteger.ONE
    private const val BOOLEAN_INDEXER_SHIFT = 32
    private val BOOLEAN_INDEXER_SIZE = BigInteger.valueOf(0b11111L) // 5 bits, 0-31
    private const val MAX_MULTIPLIER = 1 shl 4 // 16
    private const val MAX_OFFSET = 1 shl 16 // 65536

    private data class Params(val offset: BigInteger, val multiplier: BigInteger)
    private var cached: Params? = null

    private fun params(): Params {
        cached?.let { return it }
        val uid = FirebaseAuth.getInstance().currentUser?.uid
            ?: throw IllegalStateException("User not authenticated")
        val hash = sha256Hex(uid)
        val offHex = hash.substring(0, 16)
        val mulHex = hash.substring(16, 32)
        val offset = (offHex.toBigInteger(16).mod(BigInteger.valueOf(MAX_OFFSET.toLong())))
        val multiplier = (mulHex.toBigInteger(16).mod(BigInteger.valueOf(MAX_MULTIPLIER.toLong())) + BigInteger.ONE)
        val p = Params(offset, multiplier)
        cached = p
        return p
    }

    fun decryptToPrimitive(encrypted: Number): Any {
        val (offset, multiplier) = params()
        val enc = BigInteger.valueOf(encrypted.toLong())
        val encodedValue = enc.subtract(offset).divide(multiplier)
        val flag = encodedValue.and(MASK)
        val number = encodedValue.shiftRight(MASK_SIZE)

        return when (flag.toInt()) {
            0 -> number.toLong().toDouble() // INTEGER
            1 -> number.toDouble() / FLOAT_PRECISION // FLOAT
            2 -> decodeRandomBoolean(number) // BOOLEAN
            3 -> java.util.Date(number.toLong()) // DATE
            else -> number.toLong()
        }
    }

    fun decryptBoolean(encrypted: Any?): Boolean? = when (encrypted) {
        is Boolean -> encrypted
        is Number -> decryptToPrimitive(encrypted).let { it as? Boolean }
        is String -> encrypted.toBooleanStrictOrNull()
        else -> null
    }

    private fun decodeRandomBoolean(number: BigInteger): Boolean {
        val booleanIndex = number.shiftRight(BOOLEAN_INDEXER_SHIFT)
        val booleanValue = number.shiftRight(booleanIndex.toInt()).and(BOOLEAN_MASK)
        return booleanValue == BigInteger.ONE
    }

    private fun sha256Hex(input: String): String {
        val md = MessageDigest.getInstance("SHA-256")
        val hash = md.digest(input.toByteArray(Charsets.UTF_8))
        return hash.joinToString(separator = "") { b -> "%02x".format(b) }
    }
}

File: iOS/Sources/Crypt/NumericDecryptor.swift
//
//  NumericDecryptor.swift
//  GoldenUnicorn
//
//  Decrypts numbers/booleans/dates encoded by Web NumericEncryptor
//

import Foundation
import CryptoKit
import FirebaseAuth

class NumericDecryptor {
    static let shared = NumericDecryptor()
    private init() {}

    private var offset: UInt64 = 0
    private var multiplier: UInt64 = 1
    private var ready = false

    private func ensure() throws {
        if ready { return }
        guard let uid = Auth.auth().currentUser?.uid else { throw NSError(domain: "Auth", code: 401) }
        let digest = SHA256.hash(data: Data(uid.utf8))
        let hex = digest.compactMap { String(format: "%02x", $0) }.joined()
        let offHex = String(hex.prefix(16))
        let mulHex = String(hex.dropFirst(16).prefix(16))
        let MAX_MULTIPLIER: UInt64 = 1 << 4
        let MAX_OFFSET: UInt64 = 1 << 16
        offset = (UInt64(offHex, radix: 16) ?? 0) % MAX_OFFSET
        multiplier = ((UInt64(mulHex, radix: 16) ?? 0) % MAX_MULTIPLIER) + 1
        ready = true
    }

    func decryptPrimitive(_ encrypted: Any?) -> Any? {
        guard let encrypted = encrypted else { return nil }
        if let b = encrypted as? Bool { return b }
        if let n = encrypted as? NSNumber { return decryptNumber(n) }
        if let s = encrypted as? String { return Bool(s) ?? s }
        return encrypted
    }

    func decryptBool(_ value: Any?) -> Bool? {
        if let b = value as? Bool { return b }
        if let n = value as? NSNumber { return decryptNumber(n) as? Bool }
        if let s = value as? String { return Bool(s) }
        return nil
    }

    private func decryptNumber(_ n: NSNumber) -> Any? {
        do { try ensure() } catch { return n }
        var enc = n.uint64Value
        // encodedValue = (enc - offset) / multiplier
        if enc < offset { return n }
        enc = (enc - offset) / multiplier
        let flag = enc & 0b11
        let number = enc >> 2
        switch flag {
        case 0: // integer
            return Double(number)
        case 1: // float
            return Double(number) / 1000.0
        case 2: // boolean
            // booleanIndex = number >> 32; booleanValue = (number >> index) & 1
            let booleanIndex = number >> 32
            let booleanValue = (number >> booleanIndex) & 1
            return booleanValue == 1
        case 3: // date
            return Date(timeIntervalSince1970: TimeInterval(number))
        default:
            return n
        }
    }
}

